### PR TITLE
[Example] PyTorch distributed training with minGPT

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR adds a more modern distributed training example.

TODOs:
- [ ] Update the examples in our doc with this example #4468 

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

File: examples/distributed-pytorch/README.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+SkyPilot offers easy built-in environment variables to help you start distributed training easily.
+
+### Using normal `torchrun`
+
+
+The following command spawn 2 nodes with 2 L4 GPU each. 
+
+`sky launch -c train.yaml`
+
+In the [train.yaml](./train.yaml), we use `torchrun` to launch the training and set the arguments for distributed training using environment variables provided by SkyPilot.
```
Comment (romilbhardwaj) [inline examples/distributed-pytorch/README.md:26]:
```suggestion
In [train.yaml](./train.yaml), we use `torchrun` to launch the training and set the arguments for distributed training using [environment variables](https://docs.skypilot.co/en/latest/running-jobs/environment-variables.html#skypilot-environment-variables) provided by SkyPilot.
```

File: examples/distributed-pytorch/README.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+
+
+### Using `rdvz` backend
+
+`rdvz` is an alternative backend for distributed training:
+
+```
+sky launch -c train-rdzv.yaml
```
Comment (romilbhardwaj) [inline examples/distributed-pytorch/README.md:48]:
```suggestion
sky launch -c train-rdzv train-rdzv.yaml
```

File: examples/distributed-pytorch/README.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    --nproc_per_node=$SKYPILOT_NUM_GPUS_PER_NODE \
+    --master_addr=$MASTER_ADDR \
+    --master_port=8008 \
+    --node_rank=${SKYPILOT_NODE_RANK} \
+    main.py
+```
+
+
+
+### Using `rdvz` backend
```
Comment (romilbhardwaj) [inline examples/distributed-pytorch/README.md:43]:
```suggestion
### Using `rdzv` backend
```

File: examples/distributed-pytorch/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+`rdvz` is an alternative backend for distributed training:
```
Comment (romilbhardwaj) [inline examples/distributed-pytorch/README.md:45]:
```suggestion
`rdzv` is an alternative backend for distributed training:
```

File: examples/distributed-pytorch/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+```
+sky launch -c train-rdzv.yaml
+```
+
+In the [train-rdzv.yaml](./train-rdzv.yaml), we use `torchrun` to launch the training and set the arguments for distributed training using environment variables provided by SkyPilot.
```
Comment (romilbhardwaj) [inline examples/distributed-pytorch/README.md:51]:
```suggestion
In [train-rdzv.yaml](./train-rdzv.yaml), we use `torchrun` to launch the training and set the arguments for distributed training using [environment variables](https://docs.skypilot.co/en/latest/running-jobs/environment-variables.html#skypilot-environment-variables) provided by SkyPilot.
```

File: examples/distributed-pytorch/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+```
+
+
+## Scale up
+
+If you would like to scale up the training, you can simply change the resources requirement, and SkyPilot's built-in environment variables will be set automatically.
+
+For example, the following command will spawn 4 nodes with 4 L4 GPUs each.
+
+`sky launch -c train.yaml --num-nodes 2 --gpus L4:2 --cpus 8+`
```
Comment (romilbhardwaj) [inline examples/distributed-pytorch/README.md:75]: change to num nodes 4 and L4:4 
```suggestion
\```
sky launch -c train.yaml --num-nodes 4 --gpus L4:4 --cpus 8+
\```
```

**Review by romilbhardwaj — COMMENTED** [review]
Awesome, thanks @Michaelvll! Left some minor nit comments

**Comment by Michaelvll** [comment]
Thanks @romilbhardwaj for the review! Updated README

**Review by romilbhardwaj — APPROVED** [review]
Thanks @Michaelvll!

**Pull request merged**